File: Live/ios/CrowdQ Live/CrowdQ Live/BuzzController.swift
//
//  BuzzController.swift
//  CrowdQ Live
//
//  Created by Patrick Miller on 1/26/26.
//

import UIKit

class BuzzController {
    let semaphore = DispatchSemaphore(value:1)
    var workQueue = OperationQueue()
    
    init() {
        workQueue.maxConcurrentOperationCount = 1
    }
    
    func addBuzz(_ pattern : Int, _ duration : Int) {
        workQueue.addOperation {
            self.doBuzz(pattern,duration)
        }
    }
    
    private func doBuzz(_ pattern : Int, _ duration : Int) {
        var generator : UIImpactFeedbackGenerator? = nil
        switch pattern {
        case 0: // No haptic
            break
        case 1: //
            generator = UIImpactFeedbackGenerator(style: .light)
        case 2: //
            generator = UIImpactFeedbackGenerator(style: .medium)
        case 3: //
            generator = UIImpactFeedbackGenerator(style: .heavy)
        case 4: //
            generator = UIImpactFeedbackGenerator(style: .soft)
        case 5: //
            generator = UIImpactFeedbackGenerator(style: .rigid)
        default:
            generator = UIImpactFeedbackGenerator(style: .heavy)
        }
        if let generator = generator {
            generator.prepare()
            generator.impactOccurred()
        }
        Thread.sleep(forTimeInterval: Double(duration) / 1000.0)
    }
}

File: Live/android/src/BuzzController.java
package live;

import android.content.Context;
import android.os.Build;
import android.os.VibrationEffect;
import android.os.Vibrator;
import android.os.VibratorManager;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class BuzzController {
    /* Single-thread executor for queue */
    private final ExecutorService workQueue = Executors.newSingleThreadExecutor();
    private final Context context;

    public BuzzController(Context context) {
        this.context = context;
    }

    public void addBuzz(int pattern, int duration) {
        workQueue.execute(() -> doBuzz(pattern, duration));
    }

    private void doBuzz(int pattern, int duration) {
        // Equivalent to UIImpactFeedbackGenerator patterns
        long[] timings;
        int[] amplitudes;

        switch(pattern) {
            case 0: // No haptic
                return;
            case 1: // Light
                timings = new long[]{0, 50};
                amplitudes = new int[]{0, 80};
                break;
            case 2: // Medium
                timings = new long[]{0, 80};
                amplitudes = new int[]{0, 150};
                break;
            case 3: // Heavy
                timings = new long[]{0, 100};
                amplitudes = new int[]{0, 255};
                break;
            case 4: // Soft
                timings = new long[]{0, 40};
                amplitudes = new int[]{0, 60};
                break;
            case 5: // Rigid
                timings = new long[]{0, 30};
                amplitudes = new int[]{0, 220};
                break;
            default: // Heavy fallback
                timings = new long[]{0, 100};
                amplitudes = new int[]{0, 255};
                break;
        }

        Vibrator vibrator;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            VibratorManager manager = (VibratorManager) context.getSystemService(Context.VIBRATOR_MANAGER_SERVICE);
            vibrator = manager.getDefaultVibrator();
        }
        else {
            vibrator = (Vibrator) context.getSystemService(Context.VIBRATOR_SERVICE);
        }

        if (vibrator != null && vibrator.hasVibrator()) {
            VibrationEffect effect = VibrationEffect.createWaveform(timings, amplitudes, -1);
            vibrator.vibrate(effect);
        }

        // Equivalent to Thread.sleep(interval);
        try {
            Thread.sleep(duration);
        } catch (InterruptedExecution e) {
            Thread.currentThread().interrupt();
        }
    }

    public void shutdown() {
        workQueue.shutdown();
    }
}